packages/react-google-maps-api/src/components/drawing/Circle.tsx

```diff
@@ -96,71 +98,69 @@ export interface CircleProps {
   onUnmount?: (circle: google.maps.Circle) => void
 }
 
-export class Circle extends React.PureComponent<CircleProps, CircleState> {
-  static contextType = MapContext
-
-  registeredEvents: google.maps.MapsEventListener[] = []
-
-  state: CircleState = {
-    circle: null,
-  }
-
-  setCircleCallback = (): void => {
-    if (this.state.circle !== null && this.props.onLoad) {
-      this.props.onLoad(this.state.circle)
-    }
-  }
-
-  componentDidMount(): void {
-    const circle = new google.maps.Circle({
-      ...(this.props.options || {}),
-      map: this.context,
-    })
-
-    this.registeredEvents = applyUpdatersToPropsAndRegisterEvents({
-      updaterMap,
-      eventMap,
-      prevProps: {},
-      nextProps: this.props,
-      instance: circle,
-    })
-
-    this.setState(function setCircle() {
-      return {
-        circle,
-      }
-    }, this.setCircleCallback)
-  }
-
-  componentDidUpdate(prevProps: CircleProps): void {
-    if (this.state.circle !== null) {
-      unregisterEvents(this.registeredEvents)
-
-      this.registeredEvents = applyUpdatersToPropsAndRegisterEvents({
-        updaterMap,
-        eventMap,
-        prevProps,
-        nextProps: this.props,
-        instance: this.state.circle,
-      })
-    }
-  }
-
-  componentWillUnmount(): void {
-    if (this.state.circle !== null) {
-      if (this.props.onUnmount) {
-        this.props.onUnmount(this.state.circle)
+function Circle(props: CircleProps): JSX.Element {
+  const { options, onLoad, onUnmount } = props
+  const map = React.useContext(MapContext)
+  const prevProps: CircleProps = usePrevious<CircleProps>(props)
+
+  const [instance, setInstance] = React.useState<google.maps.Circle | null>(
+    null
+  )
+
+  React.useEffect(
+    function effect() {
+      if (map !== null) {
+        if (instance === null) {
+          setInstance(
+            new google.maps.Circle({
+              ...(options || {}),
+              map,
+            })
+          )
+        }
+
+        if (instance !== null) {
+          instance.setMap(map)
+
+          if (onLoad) {
+            onLoad(instance)
+          }
+        }
       }
 
-      unregisterEvents(this.registeredEvents)
+      return function cleanup() {
+        if (instance !== null) {
+          if (onUnmount) {
+            onUnmount(instance)
+          }
 
-      this.state.circle && this.state.circle.setMap(null)
-    }
-  }
+          instance.setMap(null)
+        }
+      }
+    },
+    [instance, map, options, onLoad, onUnmount]
+  )
+
+  React.useEffect(
+    function effect(): () => void {
+      const registeredEvents: google.maps.MapsEventListener[] = applyUpdatersToPropsAndRegisterEvents(
+        {
+          updaterMap,
+          eventMap,
+          prevProps,
+          nextProps: props,
+          instance,
+        }
+      )
+
+      return function cleanup(): void {
+        unregisterEvents(registeredEvents)
+      }
+    },
+    [props, instance, prevProps]
+  )
```

Comment:
Can we extract this to a custom hook used by all components? (Or maybe just the one's that have this exact functionality)

What i'm concerned about is that any time we need to make a change to this logic we'll have to change it in many components